go/chat/boxer.go

```diff
@@ -402,8 +412,276 @@ func (b *Boxer) unboxV1(ctx context.Context, boxed chat1.MessageBoxed, encryptio
 	}, nil
 }
 
-func (b *Boxer) unboxV2(ctx context.Context, msg chat1.MessageBoxed, encryptionKey *keybase1.CryptKey) (*chat1.MessageUnboxedValid, UnboxingError) {
-	return nil, NewPermanentUnboxingError(NewMessageBoxedVersionError(chat1.MessageBoxedVersion_V2))
+func (b *Boxer) unboxV2(ctx context.Context, boxed chat1.MessageBoxed, encryptionKey *keybase1.CryptKey) (*chat1.MessageUnboxedValid, UnboxingError) {
+	if boxed.ServerHeader == nil {
+		return nil, NewPermanentUnboxingError(errors.New("nil ServerHeader in MessageBoxed"))
+	}
+
+	// Check that V1-only fields are empty.
+	if len(boxed.HeaderCiphertext.E) != 0 {
+		return nil, NewPermanentUnboxingError(errors.New("populated HeaderSealed in MBV2"))
+	}
+
+	// Validate verification key against unverified sender id.
+	// Later it is asserted that the claimed and signing sender are the same.
+	senderKeyFound, senderKeyValidAtCtime, senderDeviceRevokedAt, ierr := b.ValidSenderKey(
+		ctx, boxed.ClientHeader.Sender, boxed.HeaderVerificationKey, boxed.ServerHeader.Ctime)
+	if ierr != nil {
+		return nil, ierr
+	}
+	if !senderKeyFound {
+		return nil, NewPermanentUnboxingError(libkb.NoKeyError{Msg: "sender key not found"})
+	}
+	if !senderKeyValidAtCtime {
+		return nil, NewPermanentUnboxingError(libkb.NoKeyError{Msg: "key invalid for sender at message ctime"})
+	}
```

Comment:
Comment here about how this check is kind of a standin for really solving expired keys?